lib/curl/curl.nit

```diff
@@ -486,6 +599,19 @@ class CurlResponseSuccess
 	end
 end
 
+class CurlRestResponseSuccess
+	super CurlResponseSuccessIntern
+
+	var content : nullable JsonValue = null
+
+	var status_code = 0
+
+	redef fun body_callback(line) do
+		self.content = line.to_json_value
+	end
```

Comment:
I have a few problems here and more solutions. Let's start with the problems:

1. I'm not sure about converting the result to a `JsonValue`, when you could also convert it to a `Serializable`. The input (your `CurlRestRequest`) takes a `Serializable`, I would expect the same type for the output. But I see how a `JsonValue` can be useful, so I would suggest to leave it to the user. Offer 2 methods to get the result: one returns a `JsonValue` and the other a deserialized `nullable Object`.

2. The `body_callback` method may be called more than once when receving a large JSON object. You must wait for the whole object before parsing it.

3. Deserializing a JSON string to a Nit object is pretty easy, I'm not sure it's worth adding indirection methods. But they are worth it if they also manage errors, they could be plugged into the `curl` package error management!

Solution:

1. Parse the JSON string to a `nullable Object` using a `JsonDeserializer` at the end of `CurlRestRequest::execute`, at that point we are sure that the whole JSON is downloaded.

2. If the parsing fails (`JsonDeserializer` raises errors), return a subclass of `CurlResponseFailed`, with the errors details. Return a `CurlRestResponseSuccess` on successful parsing only.

3. In `CurlRestResponseSuccess` offer 2 methods to get a parsed JSON object: `json_object` for the version deserialized to a standard Nit object, and `json_value` for the easy to explore `JsonValue` version.

~~~
class CurlRestResponseSuccess
	super CurlResponseSuccessIntern

	# Response code from the server
	var status_code = 0

	# Raw JSON string, as received from the server
	var json_string = ""

	redef fun body_callback(line) do self.json_string += line

	# Deserialized Nit object from `json_string`
	var json_object: nullable Object = null

	# `JsonValue` to access the data in `json_string`
	fun json_value: JsonValue do return new JsonValue(json_object)
end
~~~